src/resolve/resolved-dns-scope.c

```diff
@@ -1275,7 +1275,7 @@ void dns_scope_check_conflicts(DnsScope *scope, DnsPacket *p) {
 
                 /* Check for conflicts against the local cache. If so,
                  * send out an advisory query, to inform everybody */
-                r = dns_cache_check_conflicts(&scope->cache, rr, p->family, &p->sender);
+                r = dns_cache_check_conflicts(&scope->cache, rr, p->family, &p->sender, scope->link->ifname, scope->protocol);
```

Comment:
Right now resolved segfaults on this line:

```gdb
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000561504a0421c in dns_transaction_prepare (t=t@entry=0x561505aaff40, ts=<optimized out>) at ../build/src/resolve/resolved-dns-transaction.c:1707
1707	                dns_cache_prune(&t->scope->cache, t->scope->family, t->scope->link->ifname, t->scope->protocol);
```

because `t->scope->link` is `NULL`:
```gdb
(gdb) list
1702	                /* Before trying the cache, let's make sure we figured out a server to use. Should this cause
1703	                 * a change of server this might flush the cache. */
1704	                (void) dns_scope_get_dns_server(t->scope);
1705	
1706	                /* Let's then prune all outdated entries */
1707	                dns_cache_prune(&t->scope->cache, t->scope->family, t->scope->link->ifname, t->scope->protocol);
1708	
1709	                r = dns_cache_lookup(
1710	                                &t->scope->cache,
1711	                                dns_transaction_key(t),
(gdb) p t->scope
$1 = (DnsScope *) 0x561505a607b0
(gdb) p t->scope->family
$2 = 0
(gdb) p t->scope->link
$3 = (Link *) 0x0
(gdb) p t->scope->link->ifname
Cannot access memory at address 0x90
(gdb) p t->scope->protocol
$4 = DNS_PROTOCOL_DNS
```

```
# SYSTEMD_LOG_LEVEL=debug resolvectl query megasearch.net
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/dbus/system_bus_socket...
Bus n/a: changing state OPENING → AUTHENTICATING
Resolving megasearch.net (family *, interface *).
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.127 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.resolve1 path=/org/freedesktop/resolve1 interface=org.freedesktop.resolve1.Manager member=ResolveHostname cookie=2 reply_cookie=0 signature=isit error-name=n/a error-message=n/a
Got message type=error sender=org.freedesktop.DBus destination=:1.127 path=n/a interface=n/a member=n/a cookie=3 reply_cookie=2 signature=s error-name=org.freedesktop.DBus.Error.NoReply error-message=Message recipient disconnected from message bus without replying
megasearch.net: resolve call failed: Message recipient disconnected from message bus without replying
Bus n/a: changing state RUNNING → CLOSED
```

```
[  856.447404] systemd-resolved[88710]: Setting log level to debug.
[  856.447918] systemd-resolved[88710]: Sent message type=method_return sender=n/a destination=:1.126 path=n/a interface=n/a member=n/a cookie=5 reply_cookie=3 signature=n/a error-name=n/a error-message=n/a
[  864.719496] systemd-resolved[88710]: Got message type=method_call sender=:1.127 destination=org.freedesktop.resolve1 path=/org/freedesktop/resolve1 interface=org.freedesktop.resolve1.Manager member=ResolveHostname cookie=2 reply_cookie=0 signature=isit error-name=n/a e>
[  864.719997] systemd-resolved[88710]: idn2_lookup_u8: megasearch.net → megasearch.net
[  864.720073] systemd-resolved[88710]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetConnectionUnixProcessID cookie=6 reply_cookie=0 signature=s error-name=n/a error-message=n>
[  864.720203] systemd-resolved[88710]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.125 path=n/a interface=n/a member=n/a cookie=7 reply_cookie=6 signature=u error-name=n/a error-message=n/a
[  864.720321] systemd-resolved[88710]: D-Bus hostname resolution request from client PID 88769 (n/a) with UID 4294967295
[  864.720420] systemd-resolved[88710]: Looking up RR for megasearch.net IN A.
[  864.720534] systemd-resolved[88710]: Looking up RR for megasearch.net IN AAAA.
[  864.720672] systemd-resolved[88710]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=7 reply_cookie=0 signature=s error-name=n/a error-message=n/a
[  864.720804] systemd-resolved[88710]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetNameOwner cookie=8 reply_cookie=0 signature=s error-name=n/a error-message=n/a
[  864.720937] systemd-resolved[88710]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.125 path=n/a interface=n/a member=n/a cookie=9 reply_cookie=8 signature=s error-name=n/a error-message=n/a
[  864.721095] systemd-resolved[88710]: Switching to DNS server 192.168.121.1 for interface eth0.
[  864.721530] audit[88710]: ANOM_ABEND auid=4294967295 uid=978 gid=978 ses=4294967295 pid=88710 comm="3" exe="/usr/lib/systemd/systemd-resolved" sig=11 res=1
[  864.410276] kernel: show_signal_msg: 6 callbacks suppressed
[  864.410280] kernel: 3[88710]: segfault at 90 ip 0000561504a0421c sp 00007ffe62090390 error 4 in systemd-resolved[5615049e1000+55000]
[  864.410289] kernel: Code: 1b 90 03 00 48 8d 3d 16 a1 03 00 e8 de cf fd ff 48 89 c6 e9 38 ff ff ff 48 8b 3b e8 36 da fe ff 48 8b 3b 48 8b 47 18 8b 4f 08 <48> 8b 90 90 00 00 00 8b 77 0c 48 83 c7 20 e8 d7 7f fe ff 4c 8d 53
[  864.410320] kernel: audit: type=1701 audit(1625822051.757:1119): auid=4294967295 uid=978 gid=978 ses=4294967295 pid=88710 comm="3" exe="/usr/lib/systemd/systemd-resolved" sig=11 res=1
```

Reply:
*sigh*, missed the correct file, the line in question should've been https://github.com/systemd/systemd/pull/20053/files#diff-d63d6fd38d6a715e4ca052fc0fb65eda859f3822dbddffa4a87a3ee872e25eafR1707

Reply:
Relevant structures for reference:

```gdb
(gdb) p *t
$4 = {
  scope = 0x561505a607b0,
  key = 0x5615057000c0,
  bypass = 0x0,
  query_flags = 31,
  sent = 0x0,
  received = 0x0,
  answer = 0x0,
  answer_rcode = 0,
  answer_dnssec_result = _DNSSEC_RESULT_INVALID,
  answer_source = _DNS_TRANSACTION_SOURCE_INVALID,
  answer_nsec_ttl = 4294967295,
  answer_errno = 0,
  state = DNS_TRANSACTION_NULL,
  answer_query_flags = 0,
  validated_keys = 0x0,
  start_usec = 864719179,
  next_attempt_after = 0,
  timeout_event_source = 0x0,
  n_attempts = 1,
  dns_udp_fd = -1,
  dns_udp_event_source = 0x0,
  stream = 0x0,
  server = 0x0,
  current_feature_level = _DNS_SERVER_FEATURE_LEVEL_INVALID,
  clamp_feature_level_servfail = _DNS_SERVER_FEATURE_LEVEL_INVALID,
  clamp_feature_level_nxdomain = _DNS_SERVER_FEATURE_LEVEL_INVALID,
  id = 39364,
  tried_stream = false,
  initial_jitter_scheduled = false,
  initial_jitter_elapsed = false,
  probing = false,
  notify_query_candidates = 0x5615054e4a70,
  notify_query_candidates_done = 0x5615054e43e0,
  notify_zone_items = 0x0,
  notify_zone_items_done = 0x0,
  notify_transactions = 0x0,
  notify_transactions_done = 0x0,
  dnssec_transactions = 0x0,
  n_picked_servers = 0,
  block_gc = 0,
  transactions_by_scope_next = 0x0,
  transactions_by_scope_prev = 0x561505a9dc90,
  transactions_by_stream_next = 0x0,
  transactions_by_stream_prev = 0x0,
  transactions_by_key_next = 0x0,
  transactions_by_key_prev = 0x0
}
(gdb) p *(t->scope)
$5 = {
  manager = 0x5615054e39c0,
  protocol = DNS_PROTOCOL_DNS,
  family = 0,
  dnssec_mode = DNSSEC_ALLOW_DOWNGRADE,
  dns_over_tls_mode = DNS_OVER_TLS_NO,
  link = 0x0,
  cache = {
    by_key = 0x0,
    by_expiry = 0x0,
    n_hit = 0,
    n_miss = 0
  },
  zone = {
    by_key = 0x0,
    by_name = 0x0
  },
  conflict_queue = 0x0,
  conflict_event_source = 0x0,
  announce_event_source = 0x0,
  ratelimit = {
    interval = 1000000,
    burst = 1000,
    num = 0,
    begin = 0
  },
  resend_timeout = 100000,
  max_rtt = 0,
  query_candidates = 0x56150563e0c0,
  transactions_by_key = 0x5615054e4418,
  transactions = 0x561505a9dc90,
  scopes_next = 0x0,
  scopes_prev = 0x561505a84000,
  announced = false
}
```